tfjs-backend-webgpu/src/backend_webgpu.ts

```diff
@@ -573,97 +573,55 @@ export class WebGPUBackend extends KernelBackend {
-          uniformDataView.setInt32(dataViewIndex * BYTES_PER_ELEMENT, d, true);
-          dataViewIndex++;
-        });
-      } else if (array.type === 'uint32') {
-        arrayData.forEach(d => {
-          uniformDataView.setUint32(dataViewIndex * BYTES_PER_ELEMENT, d, true);
-          dataViewIndex++;
-        });
-      } else {
-        arrayData.forEach(d => {
-          uniformDataView.setFloat32(
-              dataViewIndex * BYTES_PER_ELEMENT, d, true);
-          dataViewIndex++;
-        });
-      }
-    });
-
-    return uniformDataView;
-  }
-
-  private computePadding(uniformsWithType:
-                             Array<{type: string; data: number[];}>): DataView {
+  private makeUniforms(uniformsWithType:
+                           Array<{type: string; data: number[];}>):
+      GPUBindingResource {
     let currentOffset = 0;
-    let padding = 0;
-    let dataViewIndex = 0;
-    const dimUniformsData: Array<{type: string; data: number[];}> = [];
+    // 512 is enough for current uniform size.
+    const maxUniformSize = 512;
+    const arrayBuffer = new ArrayBuffer(maxUniformSize);
     uniformsWithType.forEach((d, i) => {
       if (d.data.length === 0) {
         d.data = [1];
       }
-      // Complete std140 layout rules are documented here:
-      // tslint:disable-next-line:max-line-length
-      // https://www.khronos.org/registry/OpenGL/specs/gl/glspec45.core.pdf#page=159
+      // https://www.w3.org/TR/WGSL/#alignof
       let baseAlignment: number;
       switch (d.data.length) {
         case 0:
-          baseAlignment = 1;
+          baseAlignment = 4;
           break;
         case 1:
-          baseAlignment = 1;
+          baseAlignment = 4;
           break;
         case 2:
-          baseAlignment = 2;
+          baseAlignment = 8;
           break;
         case 3:
-          baseAlignment = 4;
+          baseAlignment = 16;
           break;
         case 4:
-          baseAlignment = 4;
+          baseAlignment = 16;
           break;
         default:
           util.assert(false, () => `Unsupported ${d.data.length}D shape`);
       }
 
-      padding = Math.ceil(currentOffset / baseAlignment) * baseAlignment -
-          currentOffset;
-      for (let p = 0; p < padding; ++p) {
-        dimUniformsData.push({type: d.type, data: [0]});
-        dataViewIndex++;
+      currentOffset = Math.ceil(currentOffset / baseAlignment) * baseAlignment;
+      if (d.type === 'int32') {
+        new Int32Array(arrayBuffer, currentOffset, d.data.length).set(d.data);
+      } else if (d.type === 'uint32') {
+        new Uint32Array(arrayBuffer, currentOffset, d.data.length).set(d.data);
+      } else {
+        new Float32Array(arrayBuffer, currentOffset, d.data.length).set(d.data);
       }
-      dimUniformsData.push({type: d.type, data: d.data});
-      dataViewIndex = dataViewIndex + d.data.length;
-      currentOffset += d.data.length + padding;
+      currentOffset += d.data.length * 4;
     });
 
```

Comment:
Does there need to `assert(currentOffset < maxUniformSize), ....`?

Reply:
I think so, too.

Reply:
@xhcao @Jiawei-Shao Please take another look. I am still not very comfortable for the assumed max uniform size 512. So I recovered partial original codes. Now it still needs two loops. One is to calculate the offsets. The other is to upload the data. Compared to original code, it 1) simplified the logic and reduced once copy, 2) use ArrayBuffer instead of DataView to allow to set a range of array.